### DISCUSSION TITLE
Espirit IPF Key does not match to Mtex ipfKeys

### DISCUSSION THREAD

**Discussion by snayir6** [mtex-toolbox/mtex#2079] [Ask Anything]
Hi Rudiger and Ralf,
I plotted one of the CTF files in Esprit software and Mtex and I noticed differences in the indexed colors. I used different IPF keys such as HSV, TSL, HKL to find out if I could match with the Espirit output, but so far I could not produce similar results.

I also used CS.porperGroup however it was usually closer to the original maps except HSV.

My questions
 1. Is there another ipfKey that can produce maps closer to the Esprit output
 2. Is there a way to change the default Mtex ipfKey after loading the maps? Right now I need to add this ipfKey.orientation2color to introduce the new ipfKey to the map

I added the figure that shows all of the different IPF maps with their keys and the Esprit output of the same CTF file. Thanks for your help in advance

Selda

Mtex links:
https://mtex-toolbox.github.io/EBSDAdvancedMaps.html

%%Here is the code to produce the maps and Keys
%%Produce indexed phase names in an e1, e2... order
e_ind=[];%%eligible phases index
for i=1:numel(ebsd.indexedPhasesId)% number of index phases
        e_ind=[e_ind i];
        eval(['e' num2str(i) '= ebsd(ebsd.mineralList(ebsd.indexedPhasesId(i)))']);
end

%% Plot Different IPF Key Maps
for i=e_ind%%1:numel(e_ind)
    vi=eval(genvarname(['e',  num2str(i)]));
    ipfKey = ipfTSLKey(vi); %%or try ipfTSLKey(vi.CS.properGroup)
    plot(vi("indexed"),ipfKey.orientation2color(vi("indexed").orientations));hold on
    title('ipfTSLKey', 'Units', 'normalized', 'Position', [0.5, 0.85, 0])
end
set(gca, 'ActivePositionProperty', 'OuterPosition');
set(gcf,'units','normalized','outerposition',[0 0 1 1]);%%Maximize figure window
set(findall(gcf,'-property','FontSize'),'FontWeight','normal','FontSize',60)
%%SAVE FILE
figname='IPFTSL';
filename=sprintf('EBSD_%s.png',figname);
saveas(gcf,fullfile(folder,filename),'png');

%% Plot Different IPF Keys 
for i=e_ind%%1:numel(e_ind)
    %i=e_ind(im);
    vi=eval(genvarname(['e',  num2str(i)]));
    ipfKey = ipfHSVKey(vi.CS.properGroup);
    plot(ipfKey)
    title('ipfTSLKey', 'Units', 'normalized', 'Position', [0.5, 0.85, 0])
    %%SAVE FILE
    set(gca,'box','off')
    set(findall(gcf,'-property','FontSize'),'FontWeight','normal','FontSize',font_size)
    figname='IPFTSL';
    filename=sprintf('IPF_Key_IPFTSL.png',figname);
    saveas(gcf,fullfile(folder,filename),'png');
    close all
end

![different_IPFKeys](https://github.com/mtex-toolbox/mtex/assets/51959635/10caa71a-dfc3-46d0-906d-abd84b186c30)

**Comment by kilir** [comment]
Hi,
you can fine-tune the HSV map until it probably mimics what you have in mind.
```
cs = crystalSymmetry('m-3m')
ck = ipfHSVKey(cs)
plot(ck)

%% change the gray gradient
nextAxis
ck = ipfHSVKey(cs)
ck.grayGradient = 0.75
plot(ck)
%% change the gray values
nextAxis
ck = ipfHSVKey(cs)
ck.grayValue = [0.1 0.7];
plot(ck)
%% change the gradient
nextAxis
ck = ipfHSVKey(cs);
ck.colorStretching = 0.8;
plot(ck)

%% move the white center
nextAxis
ck = ipfHSVKey(cs);
ck.whiteCenter = vector3d(0.5, 0.25, 0.82)
plot(ck)

%% or combine any of the above...
```
Hope that helps.

Cheers,
Rüdiger

**Reply by snayir6** [reply]
Hi Rudiger,
Thanks for your prompt reply and help, your suggestion worked. I tried following conditions to produce a customized IPF key which is closer to the Esprit output

Thanks,
Selda

%%Produced customized IPF key by changing HSH key
cs = ebsd('Austenite, fcc (New)').CS;%% or crystalSymmetry('m-3m');%% or 
ipfKey = ipfHSVKey(cs);
%%Change HSV settings
ipfKey.grayGradient =3;
ipfKey.grayValue = [0.1 1];
ipfKey.colorStretching =1.1;
%ipfKey.whiteCenter =vector3d(0.5, 0.25, 0.82);

%%ck.whiteCenter = vector3d(0.5, 0.25, 0.82);%% Do not change it if it is not necessary to do so
plot(ipfKey)
%% PLot new IPF Key for all Phases
nextAxis
%%Produce indexed phase names in an e1, e2... order
e_ind=[];%%eligible phases index
for i=1:numel(ebsd.indexedPhasesId)% number of index phases
e_ind=[e_ind i];
eval(['e' num2str(i) '= ebsd(ebsd.mineralList(ebsd.indexedPhasesId(i)))']);
end
for i=e_ind
    vi=eval(genvarname(['e',  num2str(i)]));
    ipfKey = ipfHSVKey(vi);
    %ck = ipfHSVKey(cs)
    ipfKey.grayGradient =3
    ipfKey.grayValue = [0.1 1];
    ipfKey.colorStretching =1.1;

    plot(vi("indexed"),ipfKey.orientation2color(vi("indexed").orientations));hold on
end
    %%SAVE FILE%%%%%%%%%%%%%%%%
    set(gca,'box','off')
    set(findall(gcf,'-property','FontSize'),'FontWeight','normal','FontSize',font_size)
    figname='File1';
    filename=sprintf('%s_IPF_Key_customized.png',figname);
    saveas(gcf,fullfile(folder,filename),'png');
% close all

![AMMT_DOE05_P7_316H_Nanohardness_data(3,3)_ipf_adjust_Ferrite, bcc (New)_IPF_Key_changed](https://github.com/mtex-toolbox/mtex/assets/51959635/d1dd49d0-46da-4291-85dc-835f23dd75f5)
